### PR TITLE
[SPARK-46043][SQL][FOLLOWUP] do not resolve v2 table provider with custom session catalog

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This is a follow-up of https://github.com/apache/spark/pull/43949 to fix a breaking change. Spark allows people to provide a custom session catalog, which may return custom v2 tables based on the table provider. #43949 resolves the table provider earlier than the custom session catalog, and may break custom session catalogs. This PR fixes it by not resolving table provider if custom session catalog is present.

### Why are the changes needed?

avoid breaking custom session catalogs

### Does this PR introduce _any_ user-facing change?

no, #43949 is not released yet.

### How was this patch tested?

existing tests

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala [hunk 1]
```diff
@@ -379,10 +379,6 @@ private[sql] object CatalogV2Util {
     }
   }
 
-  def loadRelation(catalog: CatalogPlugin, ident: Identifier): Option[NamedRelation] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala:382]: This is a test-only function. I found it as I was searching all the call sites of `loadTable`.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SparkAnalyzer.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.spark.sql.catalyst.QualifiedTableName
+import org.apache.spark.sql.catalyst.analysis.Analyzer
+import org.apache.spark.sql.catalyst.catalog.CatalogTable
+import org.apache.spark.sql.connector.catalog.{CatalogManager, Table, V1Table}
+import org.apache.spark.sql.connector.expressions.Transform
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Utils.{getDataSourceOptions, getTableProvider}
+
+class SparkAnalyzer(catalogManager: CatalogManager) extends Analyzer(catalogManager) {
+
+  override def resolveDataSourceV2Table(t: Table): Table = t match {
```
Comment (allisonwang-db) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SparkAnalyzer.scala:29]: I'm still confused about why we need to separate this function from loadTable. How could a custom session catalog table fail to be resolved?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SparkAnalyzer.scala:29]: a custom v2 session catalog can extend `DelegatingCatalogExtension` and take the `Table` returned by `V2SessionCatalog.loadTable` then return something else. That said, we should keep `V2SessionCatalog.loadTable` to return `V1Table` and resolve data source v2 table provider afterwards, which is after calling `CatalogV2Utils.loadTable`

**Review by allisonwang-db — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
Merged to master

Thank you @cloud-fan @allisonwang-db

**Pull request closed without merging**